Mathlib/Algebra/Module/SnakeLemma.lean

```diff
@@ -0,0 +1,223 @@
+- `SnakeLemma.exact_δ_right`: The connecting homomorphism is exact on the left.
+
+-/
+
+open LinearMap hiding id
+open Function
+
+/-!
+Suppose we have an exact commutative diagram
+```
+        K₂ -F-→ K₃
+        |       |
+        ι₂      ι₃
+        ↓       ↓
+M₁ -f₁→ M₂ -f₂→ M₃
+|       |       |
+i₁      i₂      i₃
+↓       ↓       ↓
+N₁ -g₁→ N₂ -g₂→ N₃
+|       |
+π₁      π₂
+↓       ↓
+C₁ -G-→ C₂
+
+```
+Such that `f₂` is surjective with a (set-theoretic) section `ρ`, `g₁` is injective with a
+(set-theoretic) retraction `π`, and that `ι₃` is injective and `π₁` is surjective.
+-/
+
+variable {R} [CommRing R] {M₁ M₂ M₃ N₁ N₂ N₃}
+  [AddCommGroup M₁] [Module R M₁] [AddCommGroup M₂] [Module R M₂] [AddCommGroup M₃] [Module R M₃]
+  [AddCommGroup N₁] [Module R N₁] [AddCommGroup N₂] [Module R N₂] [AddCommGroup N₃] [Module R N₃]
+  (i₁ : M₁ →ₗ[R] N₁) (i₂ : M₂ →ₗ[R] N₂) (i₃ : M₃ →ₗ[R] N₃)
+  (f₁ : M₁ →ₗ[R] M₂) (f₂ : M₂ →ₗ[R] M₃) (hf : Exact f₁ f₂)
+  (g₁ : N₁ →ₗ[R] N₂) (g₂ : N₂ →ₗ[R] N₃) (hg : Exact g₁ g₂)
+  (h₁ : g₁.comp i₁ = i₂.comp f₁) (h₂ : g₂.comp i₂ = i₃.comp f₂)
+  (σ : M₃ → M₂) (hσ : f₂ ∘ σ = id) (ρ : N₂ → N₁) (hρ : ρ ∘ g₁ = id)
+  {K₂ K₃ C₁ C₂} [AddCommGroup K₂] [Module R K₂] [AddCommGroup K₃] [Module R K₃]
+  [AddCommGroup C₁] [Module R C₁] [AddCommGroup C₂] [Module R C₂]
+  (ι₂ : K₂ →ₗ[R] M₂) (hι₂ : Exact ι₂ i₂) (ι₃ : K₃ →ₗ[R] M₃) (hι₃ : Exact ι₃ i₃)
+  (π₁ : N₁ →ₗ[R] C₁) (hπ₁ : Exact i₁ π₁) (π₂ : N₂ →ₗ[R] C₂) (hπ₂ : Exact i₂ π₂)
+
+include hg hρ h₂ hσ hι₃ in
+lemma SnakeLemma.δ_aux (x : K₃) : g₁ (ρ (i₂ (σ (ι₃ x)))) = i₂ (σ (ι₃ x)) := by
+  obtain ⟨d, hd⟩ : i₂ (σ (ι₃ x)) ∈ range g₁ := by
+    rw [← hg.linearMap_ker_eq, mem_ker, show g₂ (i₂ _) = i₃ (f₂ _) from DFunLike.congr_fun h₂ _,
+      ← @comp_apply _ _ _ f₂ σ, hσ, id_eq, ← i₃.comp_apply,
+      hι₃.linearMap_comp_eq_zero, zero_apply]
+  rw [← hd, ← ρ.comp_apply, hρ, id_eq]
+
+include hf h₁ hρ hπ₁ in
+lemma SnakeLemma.eq_of_eq (x : K₃)
+    (y₁) (hy₁ : f₂ y₁ = ι₃ x) (z₁) (hz₁ : g₁ z₁ = i₂ y₁)
+    (y₂) (hy₂ : f₂ y₂ = ι₃ x) (z₂) (hz₂ : g₁ z₂ = i₂ y₂) : π₁ z₁ = π₁ z₂ := by
+  have := sub_eq_zero.mpr (hy₁.trans hy₂.symm)
+  rw [← map_sub, hf] at this
+  obtain ⟨d, hd⟩ := this
+  rw [← eq_sub_iff_add_eq.mp hd, map_add, ← hz₂, ← sub_eq_iff_eq_add, ← map_sub,
+    ← i₂.comp_apply, ← h₁, LinearMap.comp_apply,
+    (HasLeftInverse.injective ⟨ρ, congr_fun hρ⟩).eq_iff] at hz₁
+  rw [← sub_eq_zero, ← map_sub, hz₁, hπ₁]
+  exact ⟨_, rfl⟩
+
+/--
+**Snake Lemma**
+Supppose we have an exact commutative diagram
+```
+                K₃
+                |
+                ι₃
+                ↓
+M₁ -f₁→ M₂ -f₂→ M₃
+|       |       |
+i₁      i₂      i₃
+↓       ↓       ↓
+N₁ -g₁→ N₂ -g₂→ N₃
+|
+π₁
+↓
+C₁
+
+```
+Such that `f₂` is surjective with a (set-theoretic) section `ρ`, `g₁` is injective with a
+(set-theoretic) retraction `π`,
+then the map `π₁ ∘ ρ ∘ i₂ ∘ σ ∘ ι₃` is a linear map from `K₃` to `C₁`.
+Also see `SnakeLemma.δ'` for a noncomputable version.
```

Comment:
Perhaps something like
```suggestion
Also see `SnakeLemma.δ'` for a noncomputable version that does not require an explicit section and retraction.
```
would make the difference more clear?

Reply:
Yes, thanks!